threads.go

```diff
@@ -555,42 +578,51 @@ func (t *threads) putRecord(ctx context.Context, id thread.ID, lid peer.ID, rec
 		return err
 	}
 
-	// Save the record locally
-	// Note: These get methods will return cached nodes.
-	block, err := rec.GetBlock(ctx, t)
-	if err != nil {
-		return err
-	}
-	event, ok := block.(*cbor.Event)
-	if !ok {
-		return fmt.Errorf("invalid event")
-	}
-	header, err := event.GetHeader(ctx, t, nil)
-	if err != nil {
-		return err
-	}
-	body, err := event.GetBody(ctx, t, nil)
-	if err != nil {
-		return err
-	}
-	err = t.AddMany(ctx, []format.Node{rec, event, header, body})
-	if err != nil {
-		return err
-	}
+	for i := len(unknownRecords) - 1; i >= 0; i-- {
+		r := unknownRecords[i]
+		// Save the record locally
+		// Note: These get methods will return cached nodes.
+		block, err := r.GetBlock(ctx, t)
+		if err != nil {
+			return err
+		}
+		event, ok := block.(*cbor.Event)
+		if !ok {
+			event, err = cbor.EventFromNode(block)
+			if err != nil {
+				return fmt.Errorf("invalid event: %v", err)
+			}
```

Comment:
The cast was always failing. That's why on  `!ok` I unmarshal the raw block to an `Event`.
(Again, other ** case here?).

If you take a look at deleted code, this extra work wasn't there.
I think this worked before because records that are received from gRPC are not partial. But previous records we fetch now with `r.GetBlock(..)` (thus, only with the `DAGResolver` *are partial*).
Make sense?

Reply:
Ah yeah, makes sense! Somewhat related enhancemen: It would be nice to fetch these blocks from known peers that can return the fully loaded records. As in, not from the plain `DAGResolver` (perhaps only as a last resort).

Reply:
Oh, well.. I think I clarified this further in an above big comment.
Yeah, more directed fetching is always nicer to avoid flooding other peers.

There're also a lot of paralellized fetching for new records for logs, which aren't cancelled when one of the addrs replies with information. I think there're many optimizations we can do to all this process... eventually we will have time for this.

Reply:
👍🏼